client/json.go

```diff
@@ -14,11 +14,102 @@ import (
 	"encoding/json"
 	"io"
 	"strconv"
+	"strings"
 
 	"github.com/valyala/fastjson"
 	"golang.org/x/exp/constraints"
 )
 
+// JSONPathPart represents a part of a JSON path.
+// Json path can be either a property of an object or an index of an element in an array.
+type JSONPathPart interface {
+	// Property returns the property name if the part is a property, and a boolean indicating if the part is a property.
+	Property() (string, bool)
+	// Index returns the index if the part is an index, and a boolean indicating if the part is an index.
+	Index() (uint64, bool)
+}
+
+type propPathPart string
+type indexPathPart uint64
```

Comment:
question: Why are these different types?  Why would we want them to behave differently?

Reply:
Makes sense, thanks :)

todo: Please document this reason somewhere prominent so that readers understand why this difference exists.

Reply:
+1 for documenting this please.